src/contracts/mixins/GPv2Signing.sol

```diff
@@ -307,6 +327,32 @@ abstract contract GPv2Signing {
         );
     }
 
+    /// @dev Verifies the order has been pre-signed. The signature is the
+    /// UID of the signed order.
+    ///
+    /// @param encodedSignature The pre-sign signature reprenting the order UID.
+    /// @param orderDigest The EIP-712 signing digest derived from the order
+    /// parameters.
+    /// @return owner The address of the signer.
+    function recoverPreSigner(
+        bytes calldata encodedSignature,
+        bytes32 orderDigest,
+        uint32 validTo
+    ) internal view returns (address owner) {
+        bytes calldata orderUid = encodedSignature;
+
+        bytes32 recoveredOrderDigest;
+        uint32 recoveredValidTo;
+        (recoveredOrderDigest, owner, recoveredValidTo) = orderUid
+            .extractOrderUidParams();
+
+        require(
+            recoveredOrderDigest == orderDigest && recoveredValidTo == validTo,
+            "GPv2: invalid presign signature"
+        );
+        require(preSignature[orderUid], "GPv2: order not presigned");
```

Comment:
Gotta be careful though that we cannot use a different signature of the same user to author another order.

Reply:
I think the idea would be to have `signature` be just an address and then construct the UID from it. Something along the lines of:
```solidity
bytes memory orderUid = abi.encodePacked(orderDigest, signature, validTo);
require(preSignature[orderUid], "GPv2: order not presigned");
```

> different signature of the same user to author another order

This would require hash collision though right? Otherwise, if its a different order, then `abi.encodePacked(orderDigest, signature, validTo)` would produce a different UID as `orderDigest` would be different (and is computed by the SC). Note that attempting to replay the same order would result in the settlement failing with `order already filled`.

Reply:
When I read 

> To save some calldata and a check

I thought about which check we would omit. If we were to omit `recoveredOrderDigest == orderDigest` then someone could provide a valid pre-signed hash (for some user) with an unrelated order to bypass the check and get the unrelated order executed in that user's name.

I'm sure we won't introduce this bug, but I'd still suggest to add a unit test and I will re-review the updated implementation 😄

Reply:
The idea was this: currently we take three input values from the user: `recoveredOrderDigest`, `owner`, `recoveredValidTo`. Then we check  `recoveredOrderDigest`, `recoveredValidTo` for consistency with the computed values, and use these three values to compute the UID.
Instead, we can just use the two computed values directly, without checking them against the user-provided values, which become then useless if not as another way to generate invalid orders.
The security of the system doesn't change: if there were a way to create a malicious orders with these changes, you could create a malicious order for the current PR by using the computed `orderDigest`, `validTo` to the order signature (they can be computed with public values and don't depend on any signature).

Reply:
If we don't check `recoveredOrderDigest == orderDigest` couldn't an attacker do the following:

1. Benign SC is willing to sell 1 ETH for $10k and pre-signs for it
2. Attacker submits order + matching for benign SC selling 1 ETH for $1 using the signature in 1)

If I read your proposal correctly we would recover the order digest from the signature and find it's indeed authored by the smart contract (so settlement would go through).
However, the settlement would happen with the fake values. Even though `orderDigest` does not depend on any signature, it commits to the actual order details and thus needs to be included in the signature check somehow.

Reply:
> If we don't check recoveredOrderDigest == orderDigest couldn't an attacker do the following:

We won't check it since the order UID is always built from the computed `orderDigest` (instead of passing it into the signature and then checking that it matches the computed one).

> details and thus needs to be included in the signature check somehow.

The idea is that `hash(attackersOrder) || benignScAddress || validTo` can't collide with `hash(benignScOrder) || benignScAddress || validTo` unless you find a hash collision (either order hash collision or order UID hash collision) which shouldn't be possible (if it were, Ethereum civilization would collapse as we know it) **or** `attackersOrder == benignScOrder`.

Reply:
My proposal would be (a more efficient version than) this:
```
        bytes calldata orderUid = encodedSignature;

        (, owner, ) = orderUid.extractOrderUidParams();

        bytes memory computedOrderUid = new bytes(GPv2Order.UID_LENGTH);
        computedOrderUid.packOrderUidParams(orderDigest, owner, validTo);

        require(preSignature[computedOrderUid], "GPv2: order not presigned");
```

The reason why the attack you mention would not work is that either `orderDigest` (which comes from the order decoding) is `sell 1 ETH for $10k` (and then the order is accepted as valid, but the price is right) or `orderDigest` is `sell 1 ETH for $1`, but then the computed `recomputedOrderUid` would not be the one signed by the real owner.

Reply:
I implemented @fedgiac 's proposal and added a unit test to make sure that you can't use a pre-signature for another order.